crates/red_knot/src/symbols.rs

```diff
@@ -438,8 +441,13 @@ impl SymbolTableBuilder {
 
 impl PreorderVisitor<'_> for SymbolTableBuilder {
     fn visit_expr(&mut self, expr: &ast::Expr) {
-        if let ast::Expr::Name(ast::ExprName { id, .. }) = expr {
+        if let ast::Expr::Name(ast::ExprName { id, ctx, .. }) = expr {
             self.add_symbol(id);
+            if matches!(ctx, ast::ExprContext::Store | ast::ExprContext::Del) {
+                if let Some(curdef) = self.current_definition.clone() {
```

Comment:
How expensive to we expect `Definition` becomes. Let's say we have a function definition with all its parameters (all have a name). The worst case is that this allocates a lot of new strings. Shoul definition be wrapped in an `Rc`? Or should `definitions` be interned in an `Arena` so that we can reference definitions by ID instead?

What are the cases where we have multiple definitions out of a single enclosing definition? Should we use `take` here?

Reply:
My intention is for Definition to stay quite small, because in most cases it should just hold a `TypedNodeKey` -- we don't want to copy lots of stuff from the AST when we can just reference the AST instead.

The case for shared Definition is something like `x = y = z`, or even more complex (why we need the nested visit), `(x, (y, z)) = foo`.

Ideally I might like to only reference the `Expr` on the RHS here (or even just part of it, for the unpacking case), but we have the general problem that we can't have a `TypedNodeKey` to arbitrary `Expr`.

Reply:
We can have a `TypedNodeKey` to arbitrary expressions. You can use `ExpressionRef`

Reply:
Ooh very good, I'll need to look at that in more detail.

Reply:
It looks like using `ExpressionRef` _should_ be possible (it can be cast to `AnyNodeRef`), but it will take a bit more investigation, because `ExpressionRef` doesn't impl `AstNode`, and currently `TypedNodeKey` only works for `N: AstNode`. Not going to hold up this PR on figuring this out, but I'll add a TODO to look at it later.

Reply:
Actually on second thought as I try to write the TODO, I don't even think we can / want to store just Exprs for Assignments. The RHS of an unpacking assigment may be an expression that returns a sequence type, but isn't a tuple or list literal, so we can't necessarily pick it apart to its component bits at AST level. We will need the whole LHS and the whole RHS to do the unpacking at type level, and so effectively we need the whole statement. So although there may be other cases where we'd like to reference an Expr, this isn't actually one of them.